kadmin/ank.c

```diff
@@ -308,3 +311,209 @@ add_new_key(struct add_options *opt, int argc, char **argv)
+	krb5_warn(context, ret, "invalid rotation period: %s",
+                  key_rotation_period);
+        return ret;
+    }
+
+    if (ret == 0) {
+        memset(&princ, 0, sizeof(princ));
+        princ.kvno = 1;
+        ret = krb5_parse_name(context, name, &princ_ent);
+        if (ret)
+            krb5_warn(context, ret, "krb5_parse_name");
+    }
+    if (ret != 0)
+	return ret;
+
+    /*
+     * Check that namespace has exactly one component, and prepend
+     * WELLKNOWN/HOSTBASED-NAMESPACE
+     */
+    if (krb5_principal_get_num_comp(context, princ_ent) != 2
+        || (comp0 = krb5_principal_get_comp_string(context, princ_ent, 0)) == 0
+        || (comp1 = krb5_principal_get_comp_string(context, princ_ent, 1)) == 0
+        || *comp0 == 0 || *comp1 == 0
+        || strcmp(comp0, "krbtgt") == 0)
+	krb5_warn(context, ret = EINVAL,
+                  "namespaces must have exactly two non-empty components "
+                  "like host-base principal names");
+    if (ret == 0)
+        ret = krb5_principal_set_comp_string(context, princ_ent, 2, comp0);
+    if (ret == 0)
+        ret = krb5_principal_set_comp_string(context, princ_ent, 3, comp1);
+    if (ret == 0)
+        ret = krb5_principal_set_comp_string(context, princ_ent, 0,
+                                             "WELLKNOWN");
+    if (ret == 0)
+        ret = krb5_principal_set_comp_string(context, princ_ent, 1,
+                                             HDB_WK_NAMESPACE);
+
+    /* Set up initial key rotation extension */
+    if (ret == 0) {
+        KeyRotation kr;
+        size_t size;
+
+        /* Setup key rotation metadata in a convenient way */
+        kr.flags = int2KeyRotationFlags(0);
+        kr.base_key_kvno = 1;
+        /*
+         * Avoid kvnos 0/1/2 which don't normally appear in fully created
+         * principals.
+         */
+        kr.base_kvno = 3;
+
+        /* XXX: Sanity check */
+        kr.epoch = kre;
+        kr.period = krp;
+
+        memset(&ext, 0, sizeof(ext));
+        ext.mandatory = FALSE;
+        ext.data.element =  choice_HDB_extension_data_key_rotation;
+        ext.data.u.key_rotation.len = 1;
+        ext.data.u.key_rotation.val = &kr;
+
+        ASN1_MALLOC_ENCODE(HDB_extension, buf.data, buf.length,
+                           &ext, &size, ret);
+        add_tl(&princ, KRB5_TL_EXTENSION, &buf);
+        mask |= KADM5_TL_DATA;
+    }
+
+    if (ret == 0) {
+        princ.principal = princ_ent;
+        mask |= KADM5_PRINCIPAL | KADM5_KVNO;
+
+        ret = set_entry(context, &princ, &mask,
+                        max_ticket_life, max_renewable_life,
+                        "never", "never", attributes, NSPOLICY);
+    }
+    if (ret == 0)
+        ret = edit_entry(&princ, &mask, default_ent, default_mask);
+
+    if (ret == 0)
+        ret = kstuple2etypes(&princ, &mask, nkstuple, kstuple);
+
+    /* XXX Shouldn't need a password for this */
```

Comment:
Indeed, now we don't!  Except maybe if we're talking to an older kadmind, so let's leave this for now.